torch_xla/csrc/xla_lower_util.cpp

```diff
@@ -749,4 +749,78 @@ xla::XlaOp BuildMaskedScatter(xla::XlaOp input, xla::XlaOp mask,
       scatter_dnums);
 }
 
+std::vector<xla::XlaOp> BuildAmpForeachNonFiniteCheckAndUnscale(
+    const std::vector<xla::XlaOp>& inputs, const xla::XlaOp& found_inf_float,
+    const xla::XlaOp& inv_scale) {
+  const xla::PrimitiveType origin_type =
+      XlaHelpers::ShapeOfXlaOp(found_inf_float).element_type();
+  std::vector<xla::XlaOp> found_infs;
+  xla::XlaOp one = xla::One(inputs[0].builder(), xla::PrimitiveType::PRED);
+  for (size_t i = 0; i < inputs.size(); ++i) {
+    xla::XlaOp all_finite =
+        xla::ReduceAll(xla::IsFinite(inputs[i]), one,
+                       xla::CreateScalarAndComputation(xla::PrimitiveType::PRED,
+                                                       inputs[i].builder()));
+    found_infs.push_back(xla::Not(all_finite));
+  }
+  xla::XlaOp found_inf =
+      xla::ConvertElementType(found_inf_float, xla::PrimitiveType::PRED);
+  for (size_t i = 0; i < found_infs.size(); ++i) {
+    found_inf = xla::Or(found_inf, found_infs[i]);
+  }
+  std::vector<xla::XlaOp> results;
+
+  for (size_t i = 0; i < inputs.size(); ++i) {
+    results.push_back(inputs[i] * inv_scale);
+  }
+  results.push_back(xla::ConvertElementType(found_inf, origin_type));
+  return results;
+}
+
+std::vector<xla::XlaOp> BuildAmpUpdateScale(const xla::XlaOp& growth_tracker,
+                                            const xla::XlaOp& current_scale,
+                                            const xla::XlaOp& found_inf_float,
+                                            double scale_growth_factor,
+                                            double scale_backoff_factor,
+                                            int scale_growth_interval) {
+  xla::XlaOp one = xla::One(growth_tracker.builder(), xla::PrimitiveType::S32);
+  xla::XlaOp one_float =
+      xla::One(growth_tracker.builder(), xla::PrimitiveType::F32);
+  xla::XlaOp found_inf =
+      xla::ConvertElementType(found_inf_float, xla::PrimitiveType::PRED);
+  const auto& growth_factor = XlaHelpers::ScalarValue<float>(
+      scale_growth_factor,
+      XlaHelpers::ShapeOfXlaOp(current_scale).element_type(),
+      growth_tracker.builder());
+  const auto& backoff_factor = XlaHelpers::ScalarValue<float>(
+      scale_backoff_factor,
+      XlaHelpers::ShapeOfXlaOp(current_scale).element_type(),
+      growth_tracker.builder());
+  const auto& growth_interval = XlaHelpers::ScalarValue<int>(
+      scale_growth_interval,
+      XlaHelpers::ShapeOfXlaOp(growth_tracker).element_type(),
+      growth_tracker.builder());
+
+  xla::XlaOp all_finite = xla::Not(found_inf);
+  xla::XlaOp not_achieve_interval = xla::ConvertElementType(
+      growth_interval - one - growth_tracker, xla::PrimitiveType::PRED);
+  xla::XlaOp new_growth_tracker =
+      (growth_tracker + one) *
+      ConvertElementType(all_finite * not_achieve_interval,
```

Comment:
use `xla::And` here might be a better idea, wdyt?

Reply:
Done.